[1.19.x] Custom usage animations for items

tmvkrpxl0 opened the pull request:
This is port of #8419 
I have ported changes from the original PR to 1.19.x

tmvkrpxl0 committed:
Ported original PR to 1.19.x

sciwhiz12 commented on src/main/java/net/minecraftforge/client/IForgeArmPose.java:
```diff
@@ -0,0 +1,47 @@
+ * Register one by creating a custom {@link net.minecraft.client.model.HumanoidModel.ArmPose}
+ * and returning it in {@link IClientItemExtensions#getArmPose(LivingEntity, InteractionHand, ItemStack)}.
+ */
+public interface IForgeArmPose
```
The interface nature of this class and its name may suggest that it is an extension interface for `HumanoidModel.ArmPose`. Given that the single method on it applies a transform on the arm, it may be better to name this something like `IArmPoseTransformer`.

Additionally, if there is no reasonable and usual case in which a mod might implement this interface as a class, it may be best to mark this as a `@FunctionalInterface`.

tmvkrpxl0 replied:
I'm in for marking it with `FunctionalInterface`, but does it have any downside for mods wanting to implement it as a class?

sciwhiz12 replied:
It shouldn't have any downsides, the annotation is more of a marker to indicate to users that the interface is intended to be a functional interface. (Though, the compiler is mandated by spec to emit errors if an interface annotated with `@FunctionalInterface` is not a functional interface, akin to e.g. `@Override`.)

sciwhiz12 commented on patches/minecraft/net/minecraft/client/model/HumanoidModel.java.patch:
```diff
@@ -0,0 +1,52 @@
+          return this.f_102890_;
+       }
++      // FORGE START
++      private net.minecraftforge.client.IForgeArmPose forgeArmPose;
```
This should be `final` and marked `@Nullable`, and the original constructor modified to set this field to `null`.

tmvkrpxl0 replied:
Should I move `//FORGE START` up? I'm unsure if modified original constructor should be included between `//FORGE START` and `//FORGE END`

sciwhiz12 replied:
The marker should remain where it is, as it denotes the part of the class which is added by Forge. If one wanted to mark the modification to the original constructor, they'd use `// FORGE: `, but it's not needed here since the field it assigns is inside the block.

sciwhiz12 commented on patches/minecraft/net/minecraft/client/model/HumanoidModel.java.patch:
```diff
@@ -0,0 +1,52 @@
++
++      private ArmPose(boolean twoHanded, net.minecraftforge.client.IForgeArmPose forgeArmPose) {
++         this(twoHanded);
++         this.forgeArmPose = forgeArmPose;
```
As this constructor is only invoked by mods through the `create` method, this should do a non-null check to ensure that mods cannot pass a `null` here (and subsequently trip on the later condition in `applyTransform`).

sciwhiz12 commented on src/main/java/net/minecraftforge/client/extensions/common/IClientItemExtensions.java:
```diff
@@ -57,6 +59,20 @@ default Font getFont(ItemStack stack, FontContext context)
+      * @param itemStack    The stack being held
+      * @return A custom ArmPose that can be used to define movement of the arm
+      */
+            @Nullable
```
Formatting: extra whitespace.
```suggestion
    @Nullable
```

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/entity/player/ItemUseAnimationTest.java:
```diff
@@ -0,0 +1,104 @@
+import java.util.function.Consumer;
+
+@Mod(ItemUseAnimationTest.MOD_ID)
+public class ItemUseAnimationTest
```
Please provide a javadoc for this class which explains what feature this tests, how to test it, and what the expected results are from said test.

sciwhiz12 commented on src/test/resources/META-INF/mods.toml:
```diff
@@ -240,7 +240,8 @@ modId="custom_transformtype_test"
 modId="mega_model_test"
 [[mods]]
 modId="emissive_elements_test"
-
```
Please keep the blank line.

tmvkrpxl0 replied:
So that there's 3 blank line between `# ADD ABOVE THIS LINE` and `modId="item_use_animation_test"`?

sciwhiz12 replied:
Yep, so there are only additions to this file.

tmvkrpxl0 replied:
Done

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/entity/player/ItemUseAnimationTest.java:
```diff
@@ -0,0 +1,104 @@
+
+    private static final DeferredRegister<Item> ITEMS = DeferredRegister.create(ForgeRegistries.ITEMS, MOD_ID);
+
+    private static final RegistryObject<Item> THING = ITEMS.register("thing", () -> new ThingItem(new Item.Properties().tab(CreativeModeTab.TAB_TOOLS).food(new FoodProperties.Builder().nutrition(4).build())));
```
For convenience of testing, please mark this item's food properties as `alwaysEat()`.

tmvkrpxl0 replied:
Done

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/entity/player/ItemUseAnimationTest.java:
```diff
@@ -0,0 +1,104 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2022.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.debug.entity.player;
```
Outdated license header, please run the `updateLicenses`/`licenseFormat` task.

tmvkrpxl0 replied:
Also done

sciwhiz12 commented on src/main/java/net/minecraftforge/client/IForgeArmPose.java:
```diff
@@ -0,0 +1,47 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2022.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.client;
```
Outdated license header, please run the `updateLicenses`/`licenseFormat` task.

tmvkrpxl0 replied:
Done

sciwhiz12 reviewed (changes requested):
CC @rehwinkel as the author of the original 1.18 PR, for consideration of these review comments to your PR.

tmvkrpxl0 committed:
Changed license header

Signed-off-by: tmvkrpxl0 <ironman466@naver.com>

tmvkrpxl0 committed:
Test item is now always edible

Signed-off-by: tmvkrpxl0 <ironman466@naver.com>

tmvkrpxl0 committed:
Removed white space

Co-authored-by: sciwhiz12 <sciwhiz12@gmail.com>

tmvkrpxl0 committed:
Match original file's blank line counts

Signed-off-by: tmvkrpxl0 <ironman466@naver.com>

tmvkrpxl0 committed:
Added javadoc to test class

Signed-off-by: tmvkrpxl0 <ironman466@naver.com>

tmvkrpxl0 committed:
Renamed IForgeArmPose to IArmPoseTransformer
Marked IArmPoseTransformer as FunctionalInterface
Marked HumanoidModel$ArmPose#IForgeArmPose as nullable, and set its default value to null

Signed-off-by: tmvkrpxl0 <ironman466@naver.com>

tmvkrpxl0 commented:
@sciwhiz12 All done!

sciwhiz12 commented on patches/minecraft/net/minecraft/client/model/HumanoidModel.java.patch:
```diff
@@ -0,0 +1,60 @@
+       }
++      // FORGE START
++      @javax.annotation.Nullable
++      private net.minecraftforge.client.IArmPoseTransformer forgeArmPose;
```
This should be marked `final`, and correspondingly the patched-in constructor below should be modified to _not_ call the original constructor but set the `twoHandled` field itself.

tmvkrpxl0 replied:
Ok, changed it now

tmvkrpxl0 committed:
Finalize forgeArmPose
Stop using original constructor

Signed-off-by: tmvkrpxl0 <ironman466@naver.com>

tmvkrpxl0 committed:
Applied comments from Amadornes

Signed-off-by: tmvkrpxl0 <ironman466@naver.com>

tmvkrpxl0 commented:
@amadornes I have applied comments from you

tmvkrpxl0 committed:
Changed name from partialTicks to partialTick
Added Javadoc to IClientItemExtensions#applyForgeHandTransform

Signed-off-by: tmvkrpxl0 <ironman466@naver.com>

tmvkrpxl0 committed:
Merge branch '1.19.x' into itemanim-1.19.x

tmvkrpxl0 commented on src/test/resources/META-INF/mods.toml:
```diff
@@ -243,6 +243,8 @@ modId="mega_model_test"
 [[mods]]
+modId="anvil_update_event_fix"
+[[mods]]
 modId="item_use_animation_test"
```
This is weird, it looks fine on original file https://github.com/tmvkrpxl0/MinecraftForge/blob/itemanim-1.19.x/src/test/resources/META-INF/mods.toml
But it looks weird on commit page

tmvkrpxl0 commented:
@sciwhiz12 I didn't receive any reply from amadornes yet, so If you don't mind, could you take a look at recent changes I made instead?

sciwhiz12 commented on patches/minecraft/net/minecraft/client/renderer/ItemInHandRenderer.java.patch:
```diff
@@ -24,6 +24,14 @@
+             this.m_109322_(p_109372_, p_109377_, flag2 ? ItemTransforms.TransformType.FIRST_PERSON_RIGHT_HAND : ItemTransforms.TransformType.FIRST_PERSON_LEFT_HAND, !flag2, p_109379_, p_109380_, p_109381_);
+          } else {
+             boolean flag3 = humanoidarm == HumanoidArm.RIGHT;
++            if (net.minecraftforge.client.extensions.common.IClientItemExtensions.of(p_109377_).applyForgeHandTransform(p_109379_, f_109299_.f_91074_, humanoidarm, p_109377_, p_109373_, p_109378_, p_109376_)) {} else
```
This can be simplified into a `if (!condition)` without the `else`. Additionally, it may be good to put a `// FORGE: ` marker.

tmvkrpxl0 replied:
It is to skip any other vanilla transformations and go straight into rendering if it returns true, and that's why there's `else` there
It is meant to be `else if`, combined with if statement right after it.

sciwhiz12 replied:
The same can be achieved with `if (!condition)`, because it will take the whole `if-else` chain below as the one statement guarded by the `if (!condition)`.

As in, instead of:
```java
if (condition) {} else
if (first) {
	// ...
} else if (second) {
	// ...
}
```

We can instead do:
```java
if (!condition)
if (first) {
	// ...
} else if (second) {
	// ...
}
```

The above can be reformatted to look like the following (with the braces in comments further illustrating the point:
```java
if (!condition) /* { */
	if (first) {
		// ...
	} else if (second) {
		// ...
	}
/* } */
```

tmvkrpxl0 replied:
ah I see what you mean now, thank you

sciwhiz12 commented on patches/minecraft/net/minecraft/world/item/UseAnim.java.patch:
```diff
@@ -0,0 +1,10 @@
+    SPYGLASS,
+-   TOOT_HORN;
++   TOOT_HORN,
++   CUSTOM;
```
Could you please clarify for me when this `UseAnim` value should be used by modders?

tmvkrpxl0 replied:
It is to avoid vanilla trying to "animate" items that has custom usage animation. Should I add javadoc?

sciwhiz12 replied:
Please do so, especially so it is clearly differentiated versus `UseAnim.NONE` and marked as being added by Forge.

tmvkrpxl0 committed:
Improved if statement and added javadoc

gigaherz committed:
Merge branch '1.19.x' into itemanim-1.19.x
